### PR TITLE
Move @console/internal/const.ts to @console/shared

### PR DESCRIPTION
**Pull request by jtomasek**
The goal of this (and following) PR is to
- move common package code from `@console/internal` to `@console/shared`
- gradually remove `@console/internal` imports from `@console/shared`
- greatly reduce existing cyclic dependencies

As raised here: https://github.com/openshift/console/pull/3516#discussion_r352737992, currently all modules are re-exported all the way to the top level in `@console/shared` package so the imports are simple. Since the `@console/shared` package is going to grow, we'll most likely increasingly hit the problem of name collisions. It would be good to prepare for that and find a convention for re-exporting code in `@console/shared` package.

Option 1: (used in this PR) don't export code on topmost level and import shared code as e.g. `import { MY_CONSTANT } from '@console/shared/src/constants'`.

Option 2: Create several `@console/shared-???` packages.

Option 3: Re-export only selected modules at top level (e.g. don't re-export `console-shared/src/components/formik-fields/field-types.ts`)

Option 4: Keep exporting everything at top level and find solution to the problem once it occurs.

Any other ideas? @spadgett, @vojtechszocs, @christianvogt, @andrewballantyne

### CODE REVIEW

File: frontend/packages/console-shared/src/types/pod.ts [hunk 1]
```diff
@@ -1,6 +1,6 @@
 import { DeploymentKind, K8sResourceKind, PodKind, PodPhase } from '@console/internal/module/k8s';
 import { FirehoseResult } from '@console/internal/components/utils';
-import { AllPodStatus } from '../constants';
+import { AllPodStatus } from '../constants/pod';
```
Comment (andrewballantyne) [inline frontend/packages/console-shared/src/types/pod.ts]:
```suggestion
import { AllPodStatus } from '../constants';
```
Any reason we don't want to take directly from constants?

File: frontend/packages/console-shared/src/types/resource.ts [hunk 2]
```diff
@@ -1,5 +1,5 @@
 import { K8sResourceKind, PodKind, RouteKind } from '@console/internal/module/k8s';
-import { DEPLOYMENT_STRATEGY } from '../constants';
+import { DEPLOYMENT_STRATEGY } from '../constants/resource';
```
Comment (andrewballantyne) [inline frontend/packages/console-shared/src/types/resource.ts]:
```suggestion
import { DEPLOYMENT_STRATEGY } from '../constants';
```

**Comment by jtomasek** [comment]
> I think there is an unbalance here. I like the change, but I feel it's slanted against constants (rightly so, they're probably the first to collide on a variable name alone)... but I really don't like snippets like:
> 
> ```
> import { ... } from '@console/shared';
> import { ... } from '@console/shared/src/...';
> ```
> 
> I think it would make more of an intuitive look if we dropped `console-shared/src/index.ts`. Everything had to go one step further to get their imports:

I agree, It is just that I don't want to do all changes at once, so I just started with constants and wanted to have a discussion to establish a convention to follow in subsequent PRs. If we agree on the approach, I'll continue with changing the rest of imports and remove the `src/index.ts` as last.

File: frontend/packages/console-shared/src/types/pod.ts (hunk 1 above)
Comment (jtomasek) [reply frontend/packages/console-shared/src/types/pod.ts]: I am just trying to establish a rule to consistently follow across the codebase.
When importing code inside the package, I lean towards not using the re-exports to limit potential circular dependencies. But I am happy to change that.

**Comment by christianvogt** [comment]
I suggest that we either use multiple packages in the monorepo or start making sub modules APIs inside `@console/shared` and use the `package.json` `exports` field to remap imports:
eg.
```
  "exports": {
    "./": "./src/lib"
  }
```

So that `import { ... } from '@console/shared/forms'` works without the `src` dir for a cleaner API.

We then place only the API files that re-export code into the `lib` dir so that devs cannot import something that wasn't meant for export.

**Comment by andrewballantyne** [comment]
> I agree, It is just that I don't want to do all changes at once, so I just started with constants and wanted to have a discussion to establish a convention to follow in subsequent PRs. If we agree on the approach, I'll continue with changing the rest of imports and remove the `src/index.ts` as last.

@jtomasek Today, are you able to export everything from root? If so, perhaps we do that for this change and then another PR deal with the export issue. I would prefer an all or nothing approach to the way we treat `console/shared` otherwise we could end up with a half implementation and that'll be bad for maintenance.

**Comment by jtomasek** [comment]
> We then place only the API files that re-export code into the `lib` dir so that devs cannot import something that wasn't meant for export.

This is a nice solution, however we won't be able to apply it while there are cyclic imports between @console/internal and @console/shared (which I am trying to get rid of). There are currently parts of the code which require explicitly narrowing down the imports to avoid the errors.

E.g.: Changing https://github.com/openshift/console/blob/master/frontend/packages/kubevirt-plugin/src/selectors/vm/selectors.ts#L3 to `import { createBasicLookup } from '@console/shared/src/utils';` causes following error:
```
Uncaught TypeError: Object(...) is not a function
    at Module../public/actions/ui.ts (ui.ts:55)
```

> @jtomasek Today, are you able to export everything from root? If so, perhaps we do that for this change and then another PR deal with the export issue. I would prefer an all or nothing approach to the way we treat console/shared otherwise we could end up with a half implementation and that'll be bad for maintenance.

I am able to export everything from root but I am not able to import everything using `import { ... } from '@console/shared'` as described above.

OK, I'll proceed with removing `@console/internal` dependencies from `@console/shared` and update this PR with exporting everything from root and using `... from '@console/shared'` where possible. Once that's done we can move forward with resolving the exports (unless I hit that problem along the way).

**Comment by andrewballantyne** [comment]
> OK, I'll proceed with removing `@console/internal` dependencies from `@console/shared` and update this PR with exporting everything from root and using `... from '@console/shared'` where possible. Once that's done we can move forward with resolving the exports (unless I hit that problem along the way).

If you do run into blockers, we can do it in parts. The goal of asking it in a two step (move everything THEN smart export) was to allow for us to live on one for a while (if you had to address other things). 

I think it's okay to have `@console/shared/src/...` imports if needed to get around the issue. I'd prefer they don't go much further than one deep though (in prep for the better export model that removes `src`).

**Comment by christianvogt** [comment]
At this point while `public` folder still exists, we gotta do what just works when it comes to our imports.
But we should look at establishing the folder structure as we move files to `@console/shared`

**Comment by jtomasek** [comment]
- Reverted the code to export constants from @console/shared root
- Using direct @console/shared import for constants where possible

**Comment by vojtechszocs** [comment]
> I view `@console/shared` as a dump-zone of commonality and not something special or of a specific purpose. It just houses common items for other packages.

@andrewballantyne The `@console/shared` package is for code used by 2+ other (as you call them "specific purpose") packages.

Console monorepo package type breakdown:
- Console web app (`@console/app`)
- Console plugin (e.g. `@console/dev-console` or `@console/kubevirt-plugin`)
- Console library to provide new functionality (e.g. `@console/plugin-sdk` or `@console/topology`)
- Console common code placeholder (`@console/shared`)

Web app depends on plugins and libraries. Plugins depend on libraries. Common code is currently organized in a single package.

> We then place only the API files that re-export code into the `lib` dir so that devs cannot import something that wasn't meant for export.
> 
> Reference: https://nodejs.org/docs/latest-v12.x/api/esm.html#esm_package_exports

Big :+1: on that convention. Let's have a clear and idiomatic way of expressing package exports.

I wouldn't suggest creating more "shared" packages, we just need to organize the existing `@console/shared` one in a better way. Handling imports & cyclic dependencies is inherently a code organization issue.

File: frontend/packages/console-shared/src/types/pod.ts (hunk 1 above)
Comment (vojtechszocs) [reply frontend/packages/console-shared/src/types/pod.ts]: I don't have a strong opinion here, but if we choose one way or another, let's follow that consistently.

Assuming the current preference for imports outside the shared pkg (prefer `@console/shared`), the preference for imports inside the shared pkg would better match what @andrewballantyne suggests.

But @jtomasek is right that adding more layers of indirection, i.e. re-(re-(re..))-exports add more complexity to the overall module dependency graph and therefore risk potential for cyclic dependencies.

We should first converge to one consistent way and then improve from there. I really like @christianvogt's proposal on having `lib` as import-able API which organizes the whole pkg into logical sub-systems.

File: frontend/packages/kubevirt-plugin/integration-tests/tests/utils/consts.ts [hunk 3]
```diff
@@ -1,4 +1,3 @@
-export const DASH = '-';
```
Comment (vojtechszocs) [inline frontend/packages/kubevirt-plugin/integration-tests/tests/utils/consts.ts]: :+1:

File: frontend/public/components/overview/index.tsx [hunk 4]
```diff
@@ -5,27 +5,24 @@ import * as React from 'react';
...
 import * as plugins from '../../plugins';
 import { OverviewCRD } from '@console/plugin-sdk';
-import { TransformResourceData, OverviewItem, getResourceList } from '@console/shared';
-
-// Display name for default overview group.
-// Should not be a valid label key to avoid conflicts. https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-setexport
-const DEFAULT_GROUP_NAME = 'other resources';
-
-// Interval at which metrics are retrieved and updated
-const METRICS_POLL_INTERVAL = 30 * 1000;
```
Comment (vojtechszocs) [inline frontend/public/components/overview/index.tsx]: These two constants seem to be dup'ed from `console-shared/src/constants/resource.ts` :confused:

File: frontend/packages/console-app/src/plugin.tsx [hunk 5]
```diff
@@ -1,6 +1,6 @@
 import * as React from 'react';
 import { CogsIcon } from '@patternfly/react-icons';
-import { FLAGS } from '@console/internal/const';
+import { FLAGS } from '@console/shared/src/constants';
```
[the remark this answers is not in the document]
Comment (jtomasek) [reply frontend/packages/console-app/src/plugin.tsx]: yes, exactly.

File: frontend/packages/console-shared/src/components/dashboard/utilization-card/TopConsumerPopover.tsx [hunk 6]
```diff
@@ -12,7 +12,7 @@ import {
 } from '@console/internal/components/dashboard/with-dashboard-resources';
 import { getInstantVectorStats } from '@console/internal/components/graphs/utils';
 import { featureReducerName } from '@console/internal/reducers/features';
-import { FLAGS } from '@console/internal/const';
+import { FLAGS } from '@console/shared';
```
[the remark this answers is not in the document]
Comment (jtomasek) [reply frontend/packages/console-shared/src/components/dashboard/utilization-card/TopConsumerPopover.tsx]: good catch, I'll fix this

**Comment by jtomasek** [comment]
Fixed according to comments, rebased.

**Comment by andrewballantyne** [comment]
/lgtm
/retest

I don't think any of these failures are legit... I think the system it was testing on had an issue. Going to rerun the tests to see if it clears up.

**Comment by andrewballantyne** [comment]
@jtomasek I think there is an underlying issue with the tests when this happens. I recall we have fixed it in the past by doing a rebase / force push to trick the CI tooling to reset. If you're already on master, maybe change the commit message or something to get a new git commit hash.

**Comment by andrewballantyne** [comment]
Damn... I am at a loss then. 😕

**Comment by andrewballantyne** [comment]
@jtomasek I had hoped getting the bot to fire over and over again would break through and get some success... it appears not to be the case. I think you may need to recreate your PR. Perhaps @spadgett has a better idea, but I think you just need to push the code to a new PR and let the CI train treat it as a fresh PR. 🤷‍♂ 

/hold

Stopping the bot spam.

**Comment by jtomasek** [comment]
I think I've fixed the problem. The rebase introduced more '@console/internal/const` imports...

/hold cancel

**Comment by andrewballantyne** [comment]
> I think I've fixed the problem. The rebase introduced more '@console/internal/const` imports...
> 
> /hold cancel

Was there a legit failure in there?

**Comment by jtomasek** [comment]
> Was there a legit failure in there?
yes

**Comment by andrewballantyne** [comment]
@jtomasek Do you see status moving? It looks all paused from your `retest` 5 hours ago...

**Comment by andrewballantyne** [comment]
> @jtomasek: PR needs rebase.

Damn... hopefully this is an easy rebase and we can get it in tomorrow.

**Pull request merged**